src/debugger/debugger.c

```diff
@@ -127,45 +145,105 @@ void mDebuggerRunTimeout(struct mDebugger* debugger, int32_t timeoutMs) {
 	size_t i;
 	size_t anyPaused = 0;
 
+#ifndef DISABLE_THREADING
+	struct mCoreThreadInternal* impl = debugger->threadImpl;
+#endif
+
 	switch (debugger->state) {
 	case DEBUGGER_RUNNING:
-		if (!debugger->platform->hasBreakpoints(debugger->platform)) {
-			debugger->core->runLoop(debugger->core);
-		} else {
-			debugger->core->step(debugger->core);
-			debugger->platform->checkBreakpoints(debugger->platform);
+		bool hasBreakpoints = debugger->platform->hasBreakpoints(debugger->platform);
+#ifndef DISABLE_THREADING
+		if (impl) {
```

Comment:
Almost all of these checks can be put outside of the switch statement by just doing an unlock before and lock afterwards. The only exceptions are the early exit if the thread is paused, which doesn't really seem like it belongs in this function anyway (are there any other callsites that can't do the check itself?), and the two cases at the end, which you can do an early check for before unlocking. By doing the checks external to the switch statement, you can reduce duplication of code and massively reduce the diff size at the same time.